File: paparazzi-gradle-plugin/src/test/projects/screenshot-test-kapt/processor/src/main/java/app/cash/paparazzi/processor/GreetingProcessor.java
package app.cash.paparazzi.processor;

import java.io.Writer;
import java.util.Set;
import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;
import javax.annotation.processing.SupportedSourceVersion;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.TypeElement;
import javax.tools.JavaFileObject;

@SupportedAnnotationTypes("app.cash.paparazzi.processor.GenerateGreeting")
@SupportedSourceVersion(SourceVersion.RELEASE_17)
public class GreetingProcessor extends AbstractProcessor {
  private boolean generated = false;

  @Override
  public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
    if (generated || roundEnv.getElementsAnnotatedWith(GenerateGreeting.class).isEmpty()) {
      return false;
    }
    generated = true;
    try {
      JavaFileObject file =
          processingEnv.getFiler().createSourceFile("app.cash.paparazzi.plugin.test.Greeting");
      try (Writer writer = file.openWriter()) {
        writer.write("package app.cash.paparazzi.plugin.test;\n");
        writer.write("public final class Greeting {\n");
        writer.write("  public static String text() { return \"generated-by-kapt\"; }\n");
        writer.write("}\n");
      }
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
    return true;
  }
}

File: paparazzi-gradle-plugin/src/test/projects/screenshot-test-ksp/processor/src/main/kotlin/app/cash/paparazzi/processor/GreetingProcessor.kt
package app.cash.paparazzi.processor

import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.Dependencies
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.processing.SymbolProcessor
import com.google.devtools.ksp.processing.SymbolProcessorEnvironment
import com.google.devtools.ksp.processing.SymbolProcessorProvider
import com.google.devtools.ksp.symbol.KSAnnotated

annotation class GenerateGreeting

class GreetingProcessorProvider : SymbolProcessorProvider {
  override fun create(environment: SymbolProcessorEnvironment): SymbolProcessor =
    GreetingProcessor(environment.codeGenerator)
}

class GreetingProcessor(private val codeGenerator: CodeGenerator) : SymbolProcessor {
  private var done = false

  override fun process(resolver: Resolver): List<KSAnnotated> {
    if (done) return emptyList()
    val symbols = resolver
      .getSymbolsWithAnnotation("app.cash.paparazzi.processor.GenerateGreeting")
      .toList()
    if (symbols.isEmpty()) return emptyList()
    done = true
    codeGenerator.createNewFile(
      Dependencies(false),
      "app.cash.paparazzi.plugin.test",
      "Greeting"
    ).bufferedWriter().use { writer ->
      writer.write("package app.cash.paparazzi.plugin.test\n\n")
      writer.write("object Greeting { const val TEXT: String = \"generated-by-ksp\" }\n")
    }
    return emptyList()
  }
}
